File: modules/Plugin/src/daybreak/abilitywar/game/list/mix/triplemix/TripleMix.java
package daybreak.abilitywar.game.list.mix.triplemix;

import daybreak.abilitywar.ability.AbilityBase;
import daybreak.abilitywar.ability.AbilityFactory.AbilityRegistration;
import daybreak.abilitywar.ability.AbilityManifest;
import daybreak.abilitywar.ability.AbilityManifest.Rank;
import daybreak.abilitywar.ability.AbilityManifest.Species;
import daybreak.abilitywar.ability.decorator.ActiveHandler;
import daybreak.abilitywar.ability.decorator.TargetHandler;
import daybreak.abilitywar.game.AbstractGame.GameTimer;
import daybreak.abilitywar.game.AbstractGame.Participant;
import daybreak.abilitywar.utils.base.collect.SetUnion;
import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.entity.LivingEntity;
import org.jetbrains.annotations.NotNull;

import java.lang.reflect.InvocationTargetException;
import java.util.Iterator;
import java.util.Set;
import java.util.StringJoiner;

@AbilityManifest(name = "트리플 믹스", rank = Rank.SPECIAL, species = Species.OTHERS, explain = "$(EXPLAIN)")
public class TripleMix extends AbilityBase implements ActiveHandler, TargetHandler {

	private AbilityBase first = null;
	private AbilityBase second = null;
	private AbilityBase third = null;

	private final Object EXPLAIN = new Object() {
		@Override
		public String toString() {
			final StringJoiner joiner = new StringJoiner("\n");
			joiner.add("§a---------------------------------");
			formatInfo(joiner, first);
			joiner.add("§a---------------------------------");
			formatInfo(joiner, second);
			joiner.add("§a---------------------------------");
			formatInfo(joiner, third);
			return joiner.toString();
		}

		private void formatInfo(final StringJoiner joiner, final AbilityBase ability) {
			if (ability != null) {
				joiner.add("§b" + ability.getName() + " §f[" + (ability.isRestricted() ? "§7능력 비활성화됨" : "§a능력 활성화됨") + "§f] " + ability.getRank().getRankName() + " " + ability.getSpecies().getSpeciesName());
				for (final Iterator<String> iterator = ability.getExplanation(); iterator.hasNext(); ) {
					joiner.add(ChatColor.RESET + iterator.next());
				}
			} else {
				joiner.add("§f능력이 없습니다.");
			}
		}
	};

	public TripleMix(Participant participant) {
		super(participant);
	}

	public AbilityBase getFirst() {
		return first;
	}

	public AbilityBase getSecond() {
		return second;
	}

	public AbilityBase getThird() {
		return third;
	}

	public void setAbility(final Class<? extends AbilityBase> first, final Class<? extends AbilityBase> second, final Class<? extends AbilityBase> third) throws SecurityException, InstantiationException, IllegalAccessException, IllegalArgumentException, InvocationTargetException {
		setAbility0(create(first, getParticipant()), create(second, getParticipant()), create(third, getParticipant()));
	}

	public void setAbility(final AbilityRegistration first, final AbilityRegistration second, final AbilityRegistration third) throws SecurityException, InstantiationException, IllegalAccessException, IllegalArgumentException, InvocationTargetException {
		setAbility0(create(first, getParticipant()), create(second, getParticipant()), create(third, getParticipant()));
	}

	private void setAbility0(AbilityBase abilityBase, AbilityBase abilityBase2, AbilityBase abilityBase3) {
		removeAbility();
		this.first = abilityBase;
		abilityBase.setRestricted(false);
		this.second = abilityBase2;
		abilityBase2.setRestricted(false);
		this.third = abilityBase3;
		abilityBase3.setRestricted(false);
	}

	@Override
	public boolean usesMaterial(Material material) {
		return (first != null && first.usesMaterial(material)) || (second != null && second.usesMaterial(material)) || (third != null && third.usesMaterial(material));
	}

	@Override
	public Set<GameTimer> getTimers() {
		if (hasAbility()) {
			return SetUnion.union(first.getTimers(), second.getTimers(), third.getTimers());
		}
		return super.getTimers();
	}

	@Override
	public Set<GameTimer> getRunningTimers() {
		if (hasAbility()) {
			return SetUnion.union(first.getRunningTimers(), second.getRunningTimers(), third.getRunningTimers());
		}
		return super.getRunningTimers();
	}

	public boolean hasAbility() {
		return first != null && second != null && third != null;
	}

	public void removeAbility() {
		if (hasAbility()) {
			if (first != null) {
				first.destroy();
				this.first = null;
			}
			if (second != null) {
				second.destroy();
				this.second = null;
			}
			if (third != null) {
				third.destroy();
				this.third = null;
			}
		}
	}

	@Override
	public boolean ActiveSkill(@NotNull Material material, @NotNull AbilityBase.ClickType clickType) {
		if (hasAbility()) {
			boolean abilityUsed = false;
			if (first instanceof ActiveHandler && ((ActiveHandler) first).ActiveSkill(material, clickType)) abilityUsed = true;
			if (second instanceof ActiveHandler && ((ActiveHandler) second).ActiveSkill(material, clickType)) abilityUsed = true;
			if (third instanceof ActiveHandler && ((ActiveHandler) third).ActiveSkill(material, clickType)) abilityUsed = true;
			return abilityUsed;
		}
		return false;
	}

	@Override
	public void TargetSkill(@NotNull Material material, @NotNull LivingEntity entity) {
		if (hasAbility()) {
			if (first instanceof TargetHandler) ((TargetHandler) first).TargetSkill(material, entity);
			if (second instanceof TargetHandler) ((TargetHandler) second).TargetSkill(material, entity);
			if (third instanceof TargetHandler) ((TargetHandler) third).TargetSkill(material, entity);
		}
	}

	@Override
	protected void onUpdate(Update update) {
		if (update == Update.RESTRICTION_CLEAR) {
			if (first != null) {
				first.setRestricted(false);
			}
			if (second != null) {
				second.setRestricted(false);
			}
			if (third != null) {
				third.setRestricted(false);
			}
		} else if (update == Update.RESTRICTION_SET) {
			if (first != null) {
				first.setRestricted(true);
			}
			if (second != null) {
				second.setRestricted(true);
			}
			if (third != null) {
				third.setRestricted(true);
			}
		} else if (update == Update.ABILITY_DESTROY) {
			if (first != null) {
				first.destroy();
				this.first = null;
			}
			if (second != null) {
				second.destroy();
				this.second = null;
			}
			if (third != null) {
				third.destroy();
				this.third = null;
			}
		}
	}
}

File: modules/Plugin/src/daybreak/abilitywar/game/list/mix/triplemix/TripleMix.kt
package daybreak.abilitywar.game.list.mix.triplemix

import daybreak.abilitywar.ability.AbilityBase
import daybreak.abilitywar.ability.AbilityBase.Update.ABILITY_DESTROY
import daybreak.abilitywar.ability.AbilityBase.Update.RESTRICTION_CLEAR
import daybreak.abilitywar.ability.AbilityBase.Update.RESTRICTION_SET
import daybreak.abilitywar.ability.AbilityManifest
import daybreak.abilitywar.ability.AbilityManifest.Rank.SPECIAL
import daybreak.abilitywar.ability.AbilityManifest.Species.OTHERS
import daybreak.abilitywar.ability.decorator.ActiveHandler
import daybreak.abilitywar.ability.decorator.TargetHandler
import daybreak.abilitywar.game.AbstractGame.Participant
import org.bukkit.ChatColor.RESET
import org.bukkit.Material
import org.bukkit.entity.LivingEntity
import java.lang.reflect.InvocationTargetException
import java.util.StringJoiner

@AbilityManifest(name = "트리플 믹스", rank = SPECIAL, species = OTHERS, explain = ["$(EXPLAIN)"])
class TripleMix(participant: Participant) : AbilityBase(participant), ActiveHandler, TargetHandler {
	var first: AbilityBase? = null
		private set
	var second: AbilityBase? = null
		private set
	var third: AbilityBase? = null
		private set
	private val EXPLAIN: Any = object : Any() {
		override fun toString(): String {
			val joiner = StringJoiner("\n")
			joiner.add("§a--------------------------------")
			formatInfo(joiner, first)
			joiner.add("§a--------------------------------")
			formatInfo(joiner, second)
			joiner.add("§a--------------------------------")
			formatInfo(joiner, third)
			return joiner.toString()
		}

		private fun formatInfo(joiner: StringJoiner, ability: AbilityBase?) {
			if (ability != null) {
				joiner.add("§b${ability.name} §f[${if (ability.isRestricted) "§7능력 비활성화됨" else "§a능력 활성화됨"}§f] ${ability.rank.rankName} ${ability.species.speciesName}")
				val iterator = ability.explanation
				while (iterator.hasNext()) {
					joiner.add(RESET.toString() + iterator.next())
				}
			} else {
				joiner.add("§f능력이 없습니다.")
			}
		}
	}

	@Throws(SecurityException::class, InstantiationException::class, IllegalAccessException::class, IllegalArgumentException::class, InvocationTargetException::class)
	fun setAbility(first: Class<out AbilityBase>, second: Class<out AbilityBase>, third: Class<out AbilityBase>) {
		removeAbility()
		this.first = create(first, participant)?.apply {
			this.isRestricted = false
		}
		this.second = create(second, participant)?.apply {
			this.isRestricted = false
		}
		this.third = create(third, participant)?.apply {
			this.isRestricted = false
		}
	}

	override fun usesMaterial(material: Material): Boolean {
		return first?.usesMaterial(material) ?: false || second?.usesMaterial(material) ?: false || third?.usesMaterial(material) ?: false
	}

	fun hasAbility(): Boolean {
		return first != null && second != null && third != null
	}

	fun removeAbility() {
		if (hasAbility()) {
			first?.destroy()
			first = null
			second?.destroy()
			second = null
			third?.destroy()
			third = null
		}
	}

	override fun ActiveSkill(material: Material, clickType: ClickType): Boolean {
		return if (hasAbility()) {
			var abilityUsed = false
			if (first is ActiveHandler && (first as ActiveHandler).ActiveSkill(material, clickType)) abilityUsed = true
			if (second is ActiveHandler && (second as ActiveHandler).ActiveSkill(material, clickType)) abilityUsed = true
			if (third is ActiveHandler && (third as ActiveHandler).ActiveSkill(material, clickType)) abilityUsed = true
			abilityUsed
		} else {
			false
		}
	}

	override fun TargetSkill(material: Material, entity: LivingEntity) {
		if (hasAbility()) {
			if (first is TargetHandler) (first as TargetHandler).TargetSkill(material, entity)
			if (second is TargetHandler) (second as TargetHandler).TargetSkill(material, entity)
			if (third is TargetHandler) (third as TargetHandler).TargetSkill(material, entity)
		}
	}

	override fun onUpdate(update: Update) {
		if (update == RESTRICTION_CLEAR) {
			if (hasAbility()) {
				first?.isRestricted = false
				second?.isRestricted = false
				third?.isRestricted = false
			}
		} else if (update == RESTRICTION_SET) {
			if (hasAbility()) {
				first?.isRestricted = true
				second?.isRestricted = true
				third?.isRestricted = true
			}
		} else if (update == ABILITY_DESTROY) {
			if (hasAbility()) {
				first?.destroy()
				first = null
				second?.destroy()
				second = null
				third?.destroy()
				third = null
			}
		}
	}
}